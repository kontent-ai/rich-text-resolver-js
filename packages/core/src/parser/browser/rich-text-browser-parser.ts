import { convertDomNodeAttributes, isElementNode, isRootNode, isTextNode } from "../../utils/browser-parser-utils.js";
import { getAllNewLineAndWhiteSpace, throwError } from "../../utils/common-utils.js";
import { DomNode } from "../parser-models.js";

export const parse = (input: string): DomNode[] => {
  const parser = new DOMParser();
  const sanitizedInput = input.replaceAll(getAllNewLineAndWhiteSpace, "");
  const document = parser.parseFromString(sanitizedInput, "text/html");

  return isRootNode(document) && document.body.firstChild
    ? Array.from(document.body.children).flatMap(parseInternal)
    : throwError("Cannot parse a node that is not a root node");
};

const parseInternal = (document: Node): DomNode => {
  if (isElementNode(document)) {
    return {
      tagName: document.tagName.toLowerCase(),
      attributes: document.hasAttributes() ? convertDomNodeAttributes(document.attributes) : {},
      children: document.hasChildNodes() ? Array.from(document.childNodes).flatMap(parseInternal) : [],
      type: "tag",
    };
  }

  if (isTextNode(document)) {
    return {
      content: document.nodeValue ?? "",
      type: "text",
    };
  }

  throw new Error("Unknown node");
};
