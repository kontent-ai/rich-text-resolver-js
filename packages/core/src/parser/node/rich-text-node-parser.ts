import * as NodeHtmlParser from "node-html-parser";
import { Node } from "node-html-parser";

import { getAllNewLineAndWhiteSpace, throwError } from "../../utils/common-utils.js";
import { isElementNode, isRootNode, isTextNode } from "../../utils/node-parser-utils.js";
import { DomNode } from "../parser-models.js";

export const parse = (input: string): DomNode[] => {
  const node = NodeHtmlParser.parse(input.replaceAll(getAllNewLineAndWhiteSpace, ""));

  return isRootNode(node)
    ? node.childNodes.flatMap(parseInternal)
    : throwError("Cannot parse node that is not a root");
};

const parseInternal = (node: Node): DomNode => {
  if (isElementNode(node)) {
    return {
      tagName: node.tagName.toLowerCase(),
      attributes: node.attributes,
      children: node.childNodes.flatMap(parseInternal),
      type: "tag",
    };
  }

  if (isTextNode(node)) {
    return {
      content: node.text,
      type: "text",
    };
  }

  throw new Error("Unknown node");
};
