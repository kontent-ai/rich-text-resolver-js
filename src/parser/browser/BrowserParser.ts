import { IParserEngine } from "../parser-models";

export class BrowserParser implements IParserEngine {
    private _parser: DOMParser;
    constructor() {
        this._parser = new DOMParser();
    }

    parse(html: string): Document { // unified method name for both types of parsers
        return this._parser.parseFromString(html, 'text/html');
    }
}