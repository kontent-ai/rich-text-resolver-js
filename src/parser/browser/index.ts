import { parse } from './RichTextBrowserParser';

export { parse as browserParse };
