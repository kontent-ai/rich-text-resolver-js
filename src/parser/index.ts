import * as runtimeEnvironment from "browser-or-node";
import { browserParse } from "./browser/index.js";
import { nodeParse } from "./node/index.js";

export * from "./parser-models.js";

export const parseHTML = (htmlInput: string) =>
  runtimeEnvironment.isBrowser
    ? browserParse(htmlInput)
    : nodeParse(htmlInput);
