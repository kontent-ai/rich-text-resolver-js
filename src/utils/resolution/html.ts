import {
  PortableTextHtmlComponents,
  PortableTextMarkComponent,
  PortableTextOptions,
  PortableTextTypeComponent,
  toHTML as toHTMLDefault,
} from "@portabletext/to-html";

import {
  PortableTextComponentOrItem,
  PortableTextExternalLink,
  PortableTextImage,
  PortableTextItemLink,
  PortableTextObject,
  PortableTextTable,
  PortableTextTableCell,
  PortableTextTableRow,
} from "../../index.js";

type RichTextCustomBlocks = Partial<{
  image: PortableTextTypeComponent<PortableTextImage>;
  componentOrItem: PortableTextTypeComponent<PortableTextComponentOrItem>;
  table: PortableTextTypeComponent<PortableTextTable>;
}>;

type RichTextCustomMarks = Partial<{
  contentItemLink: PortableTextMarkComponent<PortableTextItemLink>;
  link: PortableTextMarkComponent<PortableTextExternalLink>;
}>;

type RichTextHtmlComponents = Omit<PortableTextHtmlComponents, "types" | "marks"> & {
  types: PortableTextHtmlComponents["types"] & RichTextCustomBlocks;
  marks: PortableTextHtmlComponents["marks"] & RichTextCustomMarks;
};

/**
 * Converts array of portable text objects to HTML. Optionally, custom resolvers can be provided.
 *
 * This function is a wrapper around `toHTML` function from `@portabletext/to-html` package, with default resolvers for `sup` and `sub` marks added.
 *
 * @param blocks array of portable text objects
 * @param resolvers optional custom resolvers for Portable Text objects
 * @returns HTML string
 */
export const toHTML = (blocks: PortableTextObject[], resolvers?: PortableTextHtmlResolvers) => {
  const defaultComponentResolvers: PortableTextHtmlResolvers = {
    components: {
      types: {
        image: ({ value }) => resolveImage(value),
        table: ({ value }) => resolveTable(value),
      },
      marks: {
        link: ({ value, children }) => {
          // eslint-disable-next-line @typescript-eslint/no-unused-vars
          const { _key, _type, ...attributes } = value!;
          return `<a ${
            Object.entries(attributes)
              .map(([key, value]) => `${key}="${value}"`)
              .join(" ")
          }>${children}</a>`;
        },
        sup: ({ children }) => `<sup>${children}</sup>`,
        sub: ({ children }) => `<sub>${children}</sub>`,
      },
    },
  };

  const mergedComponentResolvers = {
    types: {
      ...defaultComponentResolvers.components.types,
      ...resolvers?.components.types,
    },
    marks: {
      ...defaultComponentResolvers.components.marks,
      ...resolvers?.components.marks,
    },
  };

  return toHTMLDefault(blocks, { components: mergedComponentResolvers });
};

/**
 * Extends `PortableTextOptions` type from `toHTML` package with
 * pre-defined types for resolution of Kontent.ai specific custom objects.
 */
export type PortableTextHtmlResolvers = Omit<PortableTextOptions, "components"> & {
  components: Partial<RichTextHtmlComponents>;
};

/**
 * Renders a portable text table to HTML.
 *
 * @param {PortableTextTable} table - The portable text table object to render.
 * @param {(value: PortableTextObject[]) => string} resolver - A function that resolves
 *        the content of each cell in the table.
 * @returns {string} The rendered table as an HTML string.
 */
export const resolveTable = (
  table: PortableTextTable,
  resolver: (value: PortableTextObject[]) => string = toHTML,
) => {
  const renderCell = (cell: PortableTextTableCell) => {
    const cellContent = resolver(cell.content);
    return `<td>${cellContent}</td>`;
  };

  const renderRow = (row: PortableTextTableRow) => {
    const cells = row.cells.map(renderCell);
    return `<tr>${cells.join("")}</tr>`;
  };

  const renderRows = () => table.rows.map(renderRow).join("");

  return `<table><tbody>${renderRows()}</tbody></table>`;
};

/**
 * Resolves an image object to HTML.
 *
 * @param {PortableTextImage} image - The portable text image object to be rendered.
 * @param {(image: PortableTextImage) => string} resolver - A resolver function that returns the image as an HTML string. Default implementation provided if not specified.
 * @returns {string} The resolved image as an HTML string.
 */
export const resolveImage = (
  image: PortableTextImage,
  resolver: (image: PortableTextImage) => string = toHTMLImageDefault,
): string => resolver(image);

/**
 * Provides a default resolver function for an image object to HTML. This function can be used as
 * a default argument for the `resolveImage` function.
 *
 * @param {PortableTextImage} image - The portable text image object to be rendered.
 * @returns {VueImage} An object representing the image, containing `src` and `alt` properties,
 *          and potentially other HTML attributes.
 */
export const toHTMLImageDefault = (image: PortableTextImage): string =>
  `<img src="${image.asset.url}" alt="${image.asset.alt}">`;
