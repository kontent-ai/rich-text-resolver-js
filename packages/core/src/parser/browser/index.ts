import { parse } from "../../parser/browser/rich-text-browser-parser.js";

export { parse as browserParse };
