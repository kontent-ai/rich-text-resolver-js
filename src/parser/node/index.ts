import { parse } from './rich-text-node-parser.js';

export { parse as nodeParse };
