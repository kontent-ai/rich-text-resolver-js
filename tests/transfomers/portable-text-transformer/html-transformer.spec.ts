import { Elements, ElementType } from "@kontent-ai/delivery-sdk";
import { PortableTextTypeComponentOptions } from "@portabletext/to-html";

import {
  ArbitraryTypedObject,
  nodesToPortableText,
  PortableTextBlock,
  PortableTextComponentOrItem,
  PortableTextExternalLink,
  PortableTextImage,
  PortableTextItemLink,
  PortableTextMark,
  PortableTextTable,
} from "../../../src";
import { browserParse } from "../../../src/parser/browser";
import { nodeParse } from "../../../src/parser/node";
import { PortableTextHtmlResolvers, resolveImage, toHTML } from "../../../src/utils/resolution/html";

jest.mock("short-unique-id", () => {
  return jest.fn().mockImplementation(() => {
    return {
      randomUUID: jest.fn().mockReturnValue("guid"),
    };
  });
});

type ResolverFunction<T extends ArbitraryTypedObject> = (value: T, children?: any) => string;

type CustomResolvers = {
  image?: ResolverFunction<PortableTextImage>;
  block?: ResolverFunction<PortableTextBlock>;
  table?: ResolverFunction<PortableTextTable>;
  component?: ResolverFunction<PortableTextComponentOrItem>;
  contentItemLink?: ResolverFunction<PortableTextItemLink>;
  link?: ResolverFunction<PortableTextExternalLink>;
  sup?: ResolverFunction<PortableTextMark>;
};

const customResolvers: Partial<CustomResolvers> = {
  image: (image) => `<img src="${image.asset.url}" alt="${image.asset.rel ?? ""}" height="800">`,
  sup: (_, children) => `<sup custom-attribute="value">${children}</sup>`,
};

describe("HTML transformer", () => {
  let richTextInput: Elements.RichTextElement;

  beforeEach(() => {
    richTextInput = {
      value: "<p><br></p>",
      type: ElementType.RichText,
      images: [],
      linkedItemCodenames: [],
      linkedItems: [
        {
          system: {
            id: "99e17fe7-a215-400d-813a-dc3608ee0294",
            name: "test item",
            codename: "test_item",
            language: "default",
            type: "test",
            collection: "default",
            sitemapLocations: [],
            lastModified: "2022-10-11T11:27:25.4033512Z",
            workflowStep: "published",
            workflow: "default",
          },
          elements: {
            text_element: {
              type: ElementType.Text,
              name: "text element",
              value: "random text value",
            },
          },
        },
      ],
      links: [],
      name: "dummy",
    };
  });

  const getPortableTextComponents = (
    element: Elements.RichTextElement,
    customResolvers: CustomResolvers = {},
  ): PortableTextHtmlResolvers => ({
    components: {
      types: {
        image: ({
          value,
        }) =>
          customResolvers.image
            ? customResolvers.image(value)
            : resolveImage(value),
        componentOrItem: ({
          value,
        }: PortableTextTypeComponentOptions<PortableTextComponentOrItem>) => {
          const linkedItem = element.linkedItems.find(
            (item) => item.system.codename === value.component._ref,
          );
          if (!linkedItem) return `Resolver for unknown type not implemented.`;

          switch (linkedItem.system.type) {
            case "test":
              return `<p>resolved value of text_element: <strong>${linkedItem.elements.text_element.value}</strong></p>`;
            default:
              return `Resolver for type ${linkedItem.system.type} not implemented.`;
          }
        },
      },
      marks: {
        contentItemLink: ({
          children,
          value,
        }) => `<a href="https://website.com/${value?.reference._ref}">${children}</a>`,
        sup: ({
          children,
          value,
        }) => customResolvers.sup ? customResolvers.sup(value, children) : `<sup>${children}</sup>`,
      },
    },
  });

  const transformAndCompare = (
    richTextValue: string,
    customResolvers: CustomResolvers = {},
  ) => {
    richTextInput.value = richTextValue;

    const browserTree = browserParse(richTextInput.value);
    const nodeTree = nodeParse(richTextInput.value);
    const nodePortableText = nodesToPortableText(nodeTree);
    const browserPortableText = nodesToPortableText(browserTree);
    const nodeResult = toHTML(
      nodePortableText,
      getPortableTextComponents(richTextInput, customResolvers),
    );
    const browserResult = toHTML(
      browserPortableText,
      getPortableTextComponents(richTextInput, customResolvers),
    );

    expect(nodeResult).toMatchSnapshot();
    expect(nodeResult).toEqual(browserResult);
  };

  it("builds basic portable text into HTML", () => {
    transformAndCompare(
      "<p><br></p><p>text<a href=\"http://google.com\" data-new-window=\"true\" title=\"linktitle\" target=\"_blank\" rel=\"noopener noreferrer\"><strong>link</strong></a></p><h1>heading</h1><p><br></p>",
    );
  });

  it("resolves internal link", () => {
    transformAndCompare(
      "<p><a data-item-id=\"23f71096-fa89-4f59-a3f9-970e970944ec\" href=\"\"><em>item</em></a></p>",
    );
  });

  it("resolves a linked item", () => {
    transformAndCompare(
      "<object type=\"application/kenticocloud\" data-type=\"item\" data-rel=\"link\" data-codename=\"test_item\"></object><p>text after component</p>",
    );
  });

  it("resolves a table using default fallback", () => {
    transformAndCompare(
      "<table><tbody>\n  <tr><td>Ivan</td><td>Jiri</td></tr>\n  <tr><td>Ondra</td><td>Dan</td></tr>\n</tbody></table>",
    );
  });

  it("resolves an asset", () => {
    transformAndCompare(
      "<figure data-asset-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\" data-image-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\"><img src=\"https://assets-us-01.kc-usercontent.com:443/cec32064-07dd-00ff-2101-5bde13c9e30c/3594632c-d9bb-4197-b7da-2698b0dab409/Riesachsee_Dia_1_1963_%C3%96sterreich_16k_3063.jpg\" data-asset-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\" data-image-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\" alt=\"\"></figure>",
    );
  });

  it("resolves an asset with custom resolver", () => {
    transformAndCompare(
      "<figure data-asset-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\" data-image-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\"><img src=\"https://assets-us-01.kc-usercontent.com:443/cec32064-07dd-00ff-2101-5bde13c9e30c/3594632c-d9bb-4197-b7da-2698b0dab409/Riesachsee_Dia_1_1963_%C3%96sterreich_16k_3063.jpg\" data-asset-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\" data-image-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\" alt=\"\"></figure>",
      customResolvers,
    );
  });

  it("resolves styled text with line breaks", () => {
    transformAndCompare(
      "</p>\n<p><strong>Strong text with line break<br>\nStrong text with line break</strong></p>",
    );
  });

  it("by default resolves text styled with sub and sup", () => {
    transformAndCompare(
      "<p><sub>Subscript text</sub><sup>Superscript text</sup></p>",
    );
  });

  it("resolves superscript with custom resolver", () => {
    transformAndCompare(
      "<p><sup>Superscript text</sup></p>",
      customResolvers,
    );
  });

  it("resolves a link using default fallback", () => {
    transformAndCompare(
      "<p><a href=\"https://website.com/12345\" target=\"_blank\" rel=\"noopener noreferrer\">link</a></p>",
    );
  });
});
