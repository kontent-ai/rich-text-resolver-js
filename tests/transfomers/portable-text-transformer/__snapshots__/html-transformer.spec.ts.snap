// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML transformer builds basic portable text into HTML 1`] = `"<p><br/></p><p>text<a href=http://google.com"><strong>link</strong></a></p><h1>heading</h1><p><br/></p>"`;

exports[`HTML transformer resolves a linked item 1`] = `"<p>resolved value of text_element: <strong>random text value</strong></p><p>text after component</p>"`;

exports[`HTML transformer resolves a table 1`] = `"<table><tbody><tr><td><p>Ivan</p></td><td><p>Jiri</p></td></tr><tr><td><p>Ondra</p></td><td><p>Dan</p></td></tr></tbody></table>"`;

exports[`HTML transformer resolves an asset 1`] = `"<img src="https://assets-us-01.kc-usercontent.com:443/cec32064-07dd-00ff-2101-5bde13c9e30c/3594632c-d9bb-4197-b7da-2698b0dab409/Riesachsee_Dia_1_1963_%C3%96sterreich_16k_3063.jpg" alt="">"`;

exports[`HTML transformer resolves an asset with custom resolver 1`] = `"<img src="https://assets-us-01.kc-usercontent.com:443/cec32064-07dd-00ff-2101-5bde13c9e30c/3594632c-d9bb-4197-b7da-2698b0dab409/Riesachsee_Dia_1_1963_%C3%96sterreich_16k_3063.jpg" alt="" height="800">"`;

exports[`HTML transformer resolves internal link 1`] = `"<p><a href="https://website.com/23f71096-fa89-4f59-a3f9-970e970944ec"><em>item</em></a></p>"`;
